tests/components/slack/test_notify.py

```diff
@@ -29,84 +36,397 @@
-    """Tests that a message pointer will be passed to Slack if specified."""
-    mock_client = Mock()
-    mock_client.chat_postMessage = AsyncMock()
     service = SlackNotificationService(None, mock_client, CONF_DATA)
 
-    expected_ts = "1624146685.064129"
-    await service.async_send_message("test", data={ATTR_THREAD_TS: expected_ts})
+    with pytest.raises(HomeAssistantError) as excinfo:
+        await service.async_send_message(
+            "test", target=["#nonexistent", "INVALID", "U123456789"]
+        )
+    assert excinfo.value.translation_domain == DOMAIN
+    assert excinfo.value.translation_key == "error_channel_not_found"
+    assert excinfo.value.translation_placeholders == {"channel_name": "nonexistent"}
 
-    mock_fn = mock_client.chat_postMessage
-    mock_fn.assert_called_once()
-    _, kwargs = mock_fn.call_args
-    assert kwargs["thread_ts"] == expected_ts
+    # No successful message sends due to resolution failures
+    assert mock_client.chat_postMessage.call_count == 0
+
+
+async def test_file_upload_targets(mock_client, hass_mock) -> None:
+    """Test file upload with different target types."""
+    mock_response = AsyncMock()
+    mock_response.raise_for_status = AsyncMock()
+    mock_response.read = AsyncMock(return_value=b"file content")
+
+    with patch(
+        "homeassistant.components.slack.notify.aiohttp_client.async_get_clientsession"
+    ) as mock_session:
+        mock_session.return_value.get.return_value.__aenter__.return_value = (
+            mock_response
+        )
+
+        service = SlackNotificationService(hass_mock, mock_client, CONF_DATA)
+
+        await service.async_send_message(
+            "test",
+            target=["#general", "U123456789"],
+            data={"file": {"url": "http://example.com/image.jpg"}},
+        )
+
+        # Should resolve targets and call files_upload_v2 once with all resolved channel IDs
+        assert mock_client.conversations_list.call_count == 2
+        assert mock_client.conversations_open.call_count == 1
+        assert mock_client.files_upload_v2.call_count == 2
+
+
+async def test_remote_file_not_allowed(hass_mock, mock_client) -> None:
+    """Test uploading a remote file from a non-allowed URL."""
+    hass_mock.config.is_allowed_external_url = Mock(return_value=False)
+    service = SlackNotificationService(hass_mock, mock_client, CONF_DATA)
+
+    await service.async_send_message(
+        "test", data={"file": {"url": "http://example.com/not-allowed.jpg"}}
+    )
+
+    mock_client.files_upload_v2.assert_not_called()
+
+
+async def test_remote_file_download_error(mock_client, hass_mock) -> None:
+    """Test handling of remote file download errors."""
+    with patch(
+        "homeassistant.components.slack.notify.aiohttp_client.async_get_clientsession"
+    ) as mock_session:
+        mock_session.return_value.get.side_effect = ClientError()
+
+        service = SlackNotificationService(hass_mock, mock_client, CONF_DATA)
```

Comment:
Usually we want to setup the service and test this through the normal APIs instead of using the object directly.